lnode/syncrandaddr.go

```diff
@@ -0,0 +1,301 @@
+		return nil, err
+	}
+
+	replyMsg := &pb.AddrNonce{}
+	err = proto.Unmarshal(replyBytes, replyMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return replyMsg, nil
+}
+
+// send to a neighbor a REQ_SYNC_ADDR_TXN request message, and wait for reply with timeout
+// return the reply messge RPL_SYNC_ADDR_TXN
+func (localNode *LocalNode) requestSyncAddrTxn(remoteNode *node.RemoteNode, addr []byte, nonce uint64) (*pb.Transactions, error) {
+
+	msg, err := NewReqSyncAddrTxnMessage(addr, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := localNode.SerializeMessage(msg, false)
+	if err != nil {
+		return nil, err
+	}
+
+	replyBytes, err := remoteNode.SendBytesSyncWithTimeout(buf, syncReplyTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	replyMsg := &pb.Transactions{}
+	err = proto.Unmarshal(replyBytes, replyMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Info("Sync txn by random address, get ", len(replyMsg.Transactions), "txns from neighbor")
+
+	return replyMsg, nil
+}
+
+// build a REQ_ADDR_NONCE request message
+func NewReqAddrNonceMessage() (*pb.UnsignedMessage, error) {
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_REQ_ADDR_NONCE,
+	}
+
+	return msg, nil
+
+}
+
+// build RPL_ADDR_NONCE reply message
+// reply with rand address my txn pool's with this address' nonce.
+func NewRplAddrNonceMessage(tp *pool.TxnPool) (*pb.UnsignedMessage, error) {
+	addrMap := tp.GetAddressList()
+	addrCount := len(addrMap)
+
+	var addr common.Uint160
+	var nonce uint64
+	if addrCount > 0 {
+		mapKeys := reflect.ValueOf(addrMap).MapKeys()
+
+		// get a random address
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		rn := r.Intn(addrCount)
+		addr = mapKeys[rn].Interface().(common.Uint160)
+
+		n, err := tp.GetNonceByTxnPool(addr) // here get expected nonce which is current last nonce + 1
+		if err != nil {
+			return nil, err
+		}
+		if n > 0 {
+			nonce = n - 1
+		} else {
+			nonce = 0
+		}
+	}
+
+	msgBody := &pb.AddrNonce{
+		Address: addr.ToArray(),
+		Nonce:   nonce,
+	}
+
+	buf, err := proto.Marshal(msgBody)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_RPL_ADDR_NONCE,
+		Message:     buf,
+	}
+
+	return msg, nil
+}
+
+// build REQ_SYNC_ADDR_TXN request message
+// This message includse my txn pool's txn address and maximum nonce.
+func NewReqSyncAddrTxnMessage(addr []byte, nonce uint64) (*pb.UnsignedMessage, error) {
+
+	msgBody := &pb.AddrNonce{Address: addr, Nonce: nonce}
+
+	buf, err := proto.Marshal(msgBody)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &pb.UnsignedMessage{
+		MessageType: pb.MessageType_REQ_SYNC_ADDR_TXN,
+		Message:     buf,
+	}
+
+	return msg, nil
+
+}
+
+// build RPL_SYNC_ADDR_TXN message. reply transactions which are needed to sync to the requestor.
+func NewRplSyncAddrTxnMessage(reqMsg *pb.AddrNonce, tp *pool.TxnPool) (*pb.UnsignedMessage, error) {
+
+	var addr160 common.Uint160
+	addr160.SetBytes(reqMsg.Address)
```

Comment:
Use `Uint160ParseFromBytes`, it also verifies byte array length

Reply:
Yes. modified.